### DISCUSSION TITLE
[Feature Request] 改进“Duplicates found”的功能

### DISCUSSION THREAD

**Discussion by ChenglongMa** [ChenglongMa/zoplicate#65] [Ideas]
@Geo123abc 您好，

之前您在 https://github.com/northword/zotero-format-metadata/issues/195 中提到了改进 Duplicates found 功能的意见，我想在这里跟您探讨一下。不知道您有没有时间？

目前 Zoplicate 的实现方案是：当引入的条目为重复条目时，用户会选择保留**已存在的**或是**新加入的**信息。然后将其**合并**。

这样做的结果是：

1. 新条目的信息会覆盖旧的；
2. 新条目**缺失**但旧条目中有的信息会保留；
3. 原来条目的附件、关系、Collection、Note 和 Tag 会保留并迁移到合并后的条目中；

所以，如果用户设定好了保留哪个条目后，
1. 工作流不会打断；
2. 合并后的条目会出现在当前 collection；
3. 会自动定位到当前collection中合并后的条目上；

不知道这样能不能满足您的需求？

现在已知的问题就是：以上操作最终都会有个**合并**动作，这会导致数据库中多出一行 `dc:replaces` 的记录，您有什么改进方案吗？

我的想法是：
1. 原则上我不想过多地干涉和破坏 Zotero 原有的逻辑；
2. `dc:replaces` 存在的目的，更多的是为了方便追溯，那么我想用户如果是**无意中**引入了重复的条目，那这个**追溯**也就没有必要了。所以：
   1. 在"重复条目"窗格中的合并操作，依然**保留** Zotero 默认的逻辑，即合并后添加新的 `dc:replaces` 关系。
   2. 在"发现重复条目"对话框或用户工作流不被打断的情况下，**取消** `dc:replaces` 关系的添加。

不知道这个方案符不符合您的想法？

谢谢！

PS: 我尽量保证这个插件不会太臃肿，而且也意在提升 Zotero 对重复条目管理的性能，比如，Zotero 默认每次点击“重复条目”窗格都会在数据库中添加一张临时表，这会造成很多冗余的数据。而使用插件后，软件只会在**需要**的时候更新重复条目数据（比如对条目的增删改查会触发检测），其余情况则不动。

非常感谢并且期待您的反馈！

**Comment by Geo123abc** [comment]
你好认真👍👍👍

我之前写的都是我自己的经验之谈，你对zotero数据库的认识比较深入


> 1. 新条目的信息会覆盖旧的；

这个我不太推荐，假设旧条目是A，新条目是B，合并时，

1. 如果使用条目A的信息更新补到B中，合并后的条目C的itemkey是和B一样的，虽然条目A在zotero内部的关系会被条目C继承，**但是在外部软件的关联关系可能会被破坏**，比如复制到obsidian等笔记软件中的`zotero://select/items/1_NNSYBQFJ`这样的链接；
2. 我推荐的是使用条目B的信息补充更新条目A，然后生成新的条目C，这样条目C会继承条目A的所有关联关系（zotero中的、zotero外的）
3. `dc:replaces`这个字段有什么用？我到现在还是不太清楚，可能需要查询zotero官方的手册，但是应该可以通过javascript强行改、删，（我之前探索过和这个字段相邻的`dc:relation`，这个表示的是zotero中related面板的关联条目，你可以参考一下[[Share] Add a Hyperlink to an annotation and jump to the target annotation](https://github.com/windingwind/zotero-actions-tags/discussions/296)   和  "在要关联双链的添加批注, 然后选中这个批注, 点击添加关联。这样就设置好了这些批注的关联关系。"([#](https://github.com/zzlb0224/zotero-annotation-manage#:~:text=%E5%9C%A8%E8%A6%81%E5%85%B3%E8%81%94%E5%8F%8C%E9%93%BE%E7%9A%84%E6%B7%BB%E5%8A%A0%E6%89%B9%E6%B3%A8%2C%20%E7%84%B6%E5%90%8E%E9%80%89%E4%B8%AD%E8%BF%99%E4%B8%AA%E6%89%B9%E6%B3%A8%2C%20%E7%82%B9%E5%87%BB%E6%B7%BB%E5%8A%A0%E5%85%B3%E8%81%94%E3%80%82%E8%BF%99%E6%A0%B7%E5%B0%B1%E8%AE%BE%E7%BD%AE%E5%A5%BD%E4%BA%86%E8%BF%99%E4%BA%9B%E6%89%B9%E6%B3%A8%E7%9A%84%E5%85%B3%E8%81%94%E5%85%B3%E7%B3%BB%E3%80%82 "《zzlb0224/zotero-annotation-manage: zotero-annotation-manage Zotero的笔记整理》"))

    1. 基于此我想，在导入条目B的时候，其实条目B还没有和zotero中的其它条目产生联系、也没有和你的知识体系产生联系，所以，**我推荐的是，把条目B的信息合并到条目A，然后保存为条目C，接着使用代码删除条目C的**​**​`dc:replaces`​**​**这个字段中关于条目B的（因为之前可能会在这个字段存在了一些元素）（这个要求可能有点吹毛求疵了）** ，也就相当于条目B是一个快递员，把一些信息快递过来就行了，使用它快递来的信息补充、完善zotero库中的条目（感觉有点太苛求了，见谅）
4. 在看你内容期间我在想，最完美的逻辑应该是zotero官方的，所以，尽量往官方靠拢吧

**Reply by ChenglongMa** [reply]
多谢反馈！您的意见非常好

1. 不过，首先，无论哪种情况数据库中都不会产生条目C。
2. 如果选择使用B的信息，确实会出现你说的情况，即数据库中只会保留B的itemkey, A的则被删除了，obsidian的关联关系可能会被破坏。_这个明天我试一下_。
3. 我觉得你建议的“**使用条目B的信息补充更新条目A**”，非常好，这样既更新了信息，同时保留了关联用的itemkey。
   1. 不过有个小问题就是，如果 Zotero 中已经存在了重复的条目A1, A2，那合并后必然只能留下二者其中的一个，这样也可能会破坏外部关联。_明天我看看有没有办法迁移这种关联关系。希望渺茫，因为无法控制外部软件的行为_ 🤔
4. 我也不清楚 `dc:replaces` 的意图，看代码里除了在合并的时候添加一条记录好像没有别的行为。
5. 在数据库中把条目B完全抹除是可以实现的，官方默认的就是把B移到 Trash 里，用户可以决定是否永久删除它。

总之非常感谢，我继续探索一下 😊

**Reply by Geo123abc** [reply]
我可能没说明白，条目C是条目A和条目B杂交出来的，条目C的itemkey和A的或者B的相同（取决于用户合并重复项的操作）

对于A1、A2这种问题，不在我考虑的范围内，我考虑的基础是：我的zotero库中已经查过重复了，并且没有重复条目存在，产生相同条目的时间是向zotero中再次添加条目的时候。

我猜测 dc:replaces 是一个未来规划版内容（官方提前架构好了底层数据库，但是相关功能还没开发）

我刚才突然有一个想法：
目前，在使用 zotero://select/items/1_NNSYBQFJ 定位zotero中条目的时候，好像只是使用“NNSYBQFJ”在zotero库的条目中进行搜索。
其实还可以再进一步：如果条目中搜索不到，再在条目的 dc:replaces 中搜索，这样就能解决“obsidian等外部zotero:// 这样的链接失效”的问题，具体实现方法可能：
1. 需要向官方提issue、或者贡献代码，
2. 在插件中实现这个伺服功能

（其实在交流过程中我在想，搞这么细致，有必要吗？但是可能是技术党的快了吧）

**Reply by ChenglongMa** [reply]
哈哈，英雄所见略同!

我也是想到这个方法，正要回复你，没想到你捷足先登了。

这个实现起来也非常简单，我明天试一下。

想问一下你是用什么插件和obsidian联动的？网上介绍的太多了。

说实话，这些东西给我的快乐在于：
1. 编程让人类生活更便捷，我认为这是计算机存在的唯一意义；
2. 分享让我感到快乐（"**Sharing is happiness**"，好像在哪见过😏）

**Reply by Geo123abc** [reply]
其实我笔记软件用的是 [siyuan-note](https://github.com/siyuan-note/siyuan)
ob和zotero的话话，我推荐这个 [PKM-er/obsidian-zotlit](https://github.com/PKM-er/obsidian-zotlit)

可以看下这个  [乱侃](https://ld246.com/article/1687500970083)
仅供参考

**Reply by ChenglongMa** [reply]
好的，多谢！

**Reply by ChenglongMa** [reply]
咱俩讨论半天，其实官方就是按你说的实现的😂，即如果外链找不到相应条目A，会按照 `dc:replaces` 关系去寻找B。
也许这也是 `dc:replaces` 的主要用途。

参考代码：[getItemFromURIs(uris)](https://github.com/zotero/zotero/blob/0b3eda5d82f2a6faa5e78d9ac375f352404e8ab0/chrome/content/zotero/xpcom/editorInstance.js#L1265-L1288)

**Reply by Geo123abc** [reply]
👍👍👍，那就不用考虑那么多了，
1. zotero库里面重复条目合并的时候只要按照官方的逻辑来就行
2. 新加的条目B如果和库里面的条目A重复，那就把B有用的信息保留下来，但是不保存B的itemkey到dc:replaces，毕竟新导入的条目B还没有和zotero库、你的知识库产生关联。上述只是对于条目B来自浏览器、ris等的情况。但是可能存在一种情况，就是你把新、旧两个zotero库合并的时候，或者文章共同作者分享过来的库，此时，不知道被合并的zotero条目是不是带有itemkey的信息。

我反思之后，觉得这样可能更好：
1. 任何情况下的合并，按照官方的合并逻辑就行，不需要分的太细，就是说，让zotero自己决定dc:replaces怎么处理；
2. 插件只需要提供一个选项（或者点击的按钮），实现下面的功能：在添加条目B到collection甲的时候，把条目A（以及其它和B重复的条目，如果有的话）添加到collection甲

**Reply by ChenglongMa** [reply]
Zoplicate 差不多就是这么做的：
1. 目前是调用的官方的合并方法，没有过多调整；
2. 目前默认是这样操作的，不过用户只能选择以新的或旧的条目信息为主，其他信息会合并一处
   * 这里有一点就是，其他和B重复的也就一块合并了，所以 collection甲和其他collection都只存在一个合并过的item

以下是个演示：

![Zoplicate_demo](https://github.com/ChenglongMa/zoplicate/assets/33501679/49b05cf2-444a-4dd8-ac78-a4beb923afff)

**Reply by Geo123abc** [reply]
嗯哈，讨论半天，原来是我对Zoplicate的功能不太懂啊

我忽视了一点：合并后，条目会被动带到当前collection

走远了，我可能是想要[这里的](https://github.com/northword/zotero-format-metadata/issues/195)操作 2：仅添加已有的条目到当前 collection

**Reply by ChenglongMa** [reply]
嗯，我可以添加一个选项，可以放弃添加新的B，只把已有的条目A添加到当前collection （原来的collection依然保留A的信息）

过两天新版本会发布，现在还有些小bug在处理🧐

> 但是我zotero安装了太多的插件，感觉系统有点吃不消

但这个问题暂时没法解决😓

**Reply by Geo123abc** [reply]
搓搓小手，期待

明白
